src/Microsoft.AspNetCore.Razor/Parser/TagHelpers/TagHelperParseTreeRewriter.cs

```diff
@@ -307,32 +315,98 @@ private IEnumerable<string> GetAttributeNames(Block tagBlock)
+                    var childBlock = (Block)child;
+                    childSpan = childBlock.FindFirstDescendentSpan();
 
                     if (childSpan == null)
                     {
+                        _attributeValueBuilder.Append(InvalidAttributeValueMarker);
                         continue;
                     }
+
+                    var childOffset = 0;
+
+                    if (childSpan.Symbols.Count > 0)
+                    {
+                        var potentialQuote = childSpan.Symbols[childSpan.Symbols.Count - 1] as HtmlSymbol;
+                        if (potentialQuote != null &&
+                            (potentialQuote.Type == HtmlSymbolType.DoubleQuote ||
+                            potentialQuote.Type == HtmlSymbolType.SingleQuote))
+                        {
+                            childOffset = 1;
+                        }
+                    }
+
+                    for (var j = 1; j < childBlock.Children.Count - childOffset; j++)
+                    {
+                        var valueChild = childBlock.Children[j];
+                        if (valueChild.IsBlock)
+                        {
+                            _attributeValueBuilder.Append(InvalidAttributeValueMarker);
+                        }
+                        else
+                        {
+                            var valueChildSpan = (Span)valueChild;
+                            for (var k = 0; k < valueChildSpan.Symbols.Count; k++)
+                            {
+                                _attributeValueBuilder.Append(valueChildSpan.Symbols[k].Content);
+                            }
+                        }
+                    }
                 }
                 else
                 {
                     childSpan = child as Span;
+
+                    // Entire attribute is a string
+                    for (var j = 0; j < childSpan.Symbols.Count; j++)
+                    {
+                        var htmlSymbol = (HtmlSymbol)childSpan.Symbols[j];
+                        if (htmlSymbol.Type == HtmlSymbolType.Equals)
+                        {
+                            for (j = j + 1; j < childSpan.Symbols.Count; j++)
+                            {
+                                htmlSymbol = (HtmlSymbol)childSpan.Symbols[j];
+
+                                if (htmlSymbol.Type != HtmlSymbolType.WhiteSpace &&
+                                    htmlSymbol.Type != HtmlSymbolType.NewLine &&
+                                    j != childSpan.Symbols.Count)
+                                {
+                                    int endMarker;
+                                    if (htmlSymbol.Type == HtmlSymbolType.DoubleQuote ||
+                                        htmlSymbol.Type == HtmlSymbolType.SingleQuote)
+                                    {
+                                        j++;
+                                        endMarker = childSpan.Symbols.Count - 1;
```

Comment:
When `endMarker` gets this value, outer loop will run an extra time. Suspect that's :ok: but it's very confusing to read. Triple-nesting the loops doesn't help. Suggest switching to a single loop with a couple of `bool` variables e.g. `afterEquals` controlling behaviour.